src/ophyd_async/core/_readable.py

```diff
@@ -15,9 +28,20 @@
     Callable[[ReadableChild], ReadableChild], Type["ConfigSignal"], Type["HintedSignal"]
 ]
 
+T = TypeVar("T")
+
+
+@dataclass
+class PerSignalConfig(Dict[SignalW[T], T]):
+    _signal_configuration: Dict[SignalW[Any], Any] = field(default_factory=dict)
+
+    @property
+    def signal_configuration(self) -> Dict[SignalW[Any], Any]:
+        return self._signal_configuration
```

Comment:
Just found out that type checking is currently disabled, I'm in the process of restoring it, but for now I will be the type checker...

The `signal_configuration` and dataclassness of this class aren't actually being used, you're actually storing everything in the dict that is `self`. Using a plain `dict` would be the same as what you have here. However, we want to support type checkers. If you replace with:
```python
class PerSignalConfig(Dict[SignalW, Any]):
    def __setitem__(self, signal: SignalW[T], value: T):
        super().__setitem__(signal, value)
    def __getitem__(self, signal: SignalW[T]) -> T:
        return super().__getitem__[signal]
```
Then we are effectively narrowing the type information just for the type checker. If we wanted to be even cleverer we could be clear that this is just for type checkers:
```python
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    # Narrow getitem and setitem so only Signals are valid
    class PerSignalConfig(Dict[SignalW, Any]):
        def __setitem__(self, signal: SignalW[T], value: T):
            super().__setitem__(signal, value)
        def __getitem__(self, signal: SignalW[T]) -> T:
            return super().__getitem__[signal]
else:
    # At runtime this is effectively the same as a dict, so might as well use one
    PerSignalConfig = dict
```

However I don't think it's worth the extra fluff for a few microseconds runtime performance.

Reply:
Ok I made the change without the additional fluff.